harness/determined/tensorboard/fetchers/s3.py

```diff
@@ -0,0 +1,88 @@
+import datetime
+import os
+import urllib.parse
+from typing import Any, Dict, Generator, List, Tuple
+
+from determined.common._logging import get_named_logger
+
+from .base import Fetcher
+
+logger = get_named_logger(__name__)
+
+
+class S3Fetcher(Fetcher):
+    def __init__(
+        self, storage_config: Dict[str, Any], storage_paths: List[str], local_dir: str
+    ) -> None:
+        import boto3
+
+        from determined.common.storage import boto3_credential_manager
+
+        boto3_credential_manager.initialize_boto3_credential_providers()
+        self.s3 = boto3.resource(
+            "s3",
+            endpoint_url=storage_config.get("endpoint_url", None),
+            aws_access_key_id=storage_config.get("access_key", None),
+            aws_secret_access_key=storage_config.get("secret_key", None),
+        )
+        self.client = self.s3.meta.client
+        self.bucket_name = str(storage_config["bucket"])
+
+        self.local_dir = local_dir
+        self.storage_paths = storage_paths
+        self._file_records = {}  # type: Dict[str, datetime.datetime]
+
+    def _find_keys(self, prefix: str) -> Generator[Tuple[str, datetime.datetime], None, None]:
+        logger.debug(f"Listing keys in bucket '{self.bucket_name}' with prefix '{prefix}'")
+
+        prefix = urllib.parse.urlparse(prefix).path.lstrip("/")
+
+        list_objs_args = {}  # type: Dict[str, Any]
+        while True:
+            list_objs_ret = self.client.list_objects_v2(
+                Bucket=self.bucket_name,
+                Prefix=prefix,
+                **list_objs_args,
+            )
+            logger.debug(f"list_objects_v2 response dict: {list_objs_ret}")
```

Comment:
Is this going to become very expensive on large experiments?

I wonder if we could save tensorflow files with timestamps in the filename and ignore anything that was too old using the `start_from` parameter or whatever it's called.

Just a though, not in scope for this PR I think.

Reply:
> Is this going to become very expensive on large experiments?
I am not sure what you mean by expensive. Do you mean for situation where there are ~10K files, we will do 10 requests per second, to list files?

So, you are thinking that at some point after the initial listing of files we then start giving a time offset to the `list_objects_v2()` call so that we only see recently updated files? This is a super good idea actually!

I am curious if Tensorboard does this already.

> not in scope for this PR I think
Cool. Will do it at some point in the future.

Reply:
> Do you mean for situation where there are ~10K files, we will do 10 requests per second, to list files?

Yes.  If each file were 50 bytes long (a low estimate given that there's a uuid in the mix) then that would be 1MB/s (8Mbps), which is probably on the order of streaming an hd movie, just to sit there doing nothing.  It's clearly an extreme example, but the point is that this strategy does not scale well.

Reply:
Do you want me to change and test this before this goes in? Or can I commit to do it after this lands? I might need to change the tests.

Reply:
Um, I'm actually not 100% certain how you would fix it.  Maybe backoff on the frequency of updates as they start to grow very large?

Anyway, no I think it's probably ok to land as-is.  But I don't feel totally confident... perhaps get another opinion, from @stoksc or @ioga?

Reply:
it's hard to judge without knowing the actual usage patterns. 
- ship it today but add more logging to catch the "expensive" cases? i.e. can we add `info` level logging for "large" responses here so we could catch it when it happens?
- strategy-wise - maybe we could backoff and poll more rarely when there're no changes to reduce the idle traffic?
- would more advanced solutions around s3 events -> SNS/SQS be viable?

Reply:
I am not sure how/where we would get the info logs for this?

I will commit to looking into investigating if:
1. Tensorboard files are only written to once and not updated (this isn't clear to me right now).
2. Dig into source for `tensorboard/backend/event_processing/data_ingester.py` and see what they do there.
3. Assuming only new files are ever written, that we rename the files so that they are sortable for use with the `StartAfter` flag.
4. Look for corresponding options in the other storage platforms we support.

Also, @rb-determined-ai, this is another example where a generic storage interface would suffer in performance :)

Reply:
I'd say:

2. Good idea, but IMO only useful if you find out that they're doing something better than us.  If what they do is as naive as us, then that isn't really good enough, since we are embarking on this mission because we're sick of dealing with tensorboard not behaving very well in corner case (that is, we really need to be better)
3. Not a good idea.  Tensorboard fetchres should be read-only.

That SNS/SQS that Ilia mentioned would be a sweet addition.  I spent some time and arrived poking around [this](https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html#supported-notification-event-types) documentation, and I don't think that would really be meaningful for us to support any time soon.

Also, @varlogtim, why did you mark this comment thread as "resolved" when it is clearly still active?

> I am not sure how/where we would get the info logs for this?

No, he means you should be logging fetches which become very large at the info level, so it's more obvious if we have a problem.   Maybe just log any time we do >1 fetch per second?